submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,415 @@
+in the cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+This could be achieved using either:
+
+- Different cable driver implementations or
+- The same cable driver implementation with differing configurations.
+
+For other use cases, it makes sense to offer finer-grained connectivity options. One such use case is a client-server
+scenario where multiple client clusters need to access a server cluster but the user does not want the client clusters
+to connect to one another. Finer-grained connectivity could also improve overall scalability where full mesh is not
+required.
+
+## Proposal
+
+This enhancement proposes a mechanism to allow users to specify their intent as to which clusters to connect and how to
+connect them. The default would still be a full mesh with a homogeneous cable driver.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect is eventually
+realized into the cluster's desired state. To extend the Kubernetes API, one could either use the generic Kubernetes
+`ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each but a CRD seems more suitable for
+this enhancement.
+
+This enhancement will propose a single API for finer-grained connectivity and multiple cable driver support.
+
+### Finer-grained connectivity and multiple cable driver API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectionPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec ClusterConnectionPolicy `json:"spec"`
+}
+
+type ClusterConnectionPolicySpec struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    // Not used in combination with ClusterSelector.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    // Not used in combination with ClusterSelector.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    // ClusterSelector identifies the Cluster resources representing a subset of clusters.
+    // Not used in combination with LeftClusterSelector or RightClusterSelector.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    ClusterSelector metav1.LabelSelector `json:"clusterSelector,omitempty"`
+
+    // Topology identifies the connection topology for the cable ends: full-mesh/point-to-point/client-server.
+    Topology `json:"topology,omitempty"`
```

Comment:
OK - I see. That would come into play if the user wants everything else not in that server-client sub-topology to connect by default. Either way they'd need another policy but with `topology` it would simply be:

```
spec:
  connectionPolicy: connect
```

instead of:

```
spec:
  leftClusterSelector:
    database-role: client
  rightClusterSelector:
    database-role: client
  connectionPolicy: no-connect
```

Reply:
yeah that's was the intention

Reply:
Let's consider the client-server use case. I have a `server` cluster along with `client1` and `client2` and another `cluster3`.  So I add a policy (or policies) to connect the the clients with the server but disallows `client1` <-> `client2`.  With your  proposal the default/fallback policy is `connect`. So the cluster connectivity would be:

  `client1` <-> `server`
  `client2` <-> `server`
  `cluster3` <-> `server`
  `cluster3` <-> `client1`
  `cluster3` <-> `client2`

The policies only disallow `client1` <-> `client2`. The 2 clients can connect to `cluster3`. Also `cluster3` is able to connect to the `server`.  This doesn't make sense to me. It **does** make sense if the default policy is `no-connect`, then you get the desired connectivity of:

  `client1` <-> `server`
  `client2` <-> `server`

This is why the original proposal took the strictly opt-in approach. Otherwise with the `connect` fallback policy, you may not be really restricting much of anything.

Reply:
@tpantelis the intention in the proposal was that the default policy is configurable. It can be connect or it can be no connect. Where none is specified the default policy will be to connect (which is keeping in line with what Submariner has done historically)...
There is some onus on the user to use with care. But this comes through examples and guides to help them reach a desirable configuration

Reply:
yeah - the default policy may not make sense depending on what the user wants to do. We could make it simpler by not having a connect or not connect choice as in the original proposal. Again the thinking/assumption is that a user would want to pare down the connectivity in which case the `no-connect` default policy makes sense - this was based on the 2 use cases that we thought of, client-server and scalability. As mentioned before, your proposal extends it to facilitate a case where the user only wants to disallow connectivity between a small number of clusters in which case a `connect` default policy makes that easier. But I don't know if there's a real use case for that.